Split IDEInspectionCallbacks into CodeCompletionCallbacks and DoneParsingCallback

ahoppen opened the pull request:
Cursor info only cares about the `doneParsing` callback and not about all the `complete` functions that are now defined in `CodeCompletionCallbacks`. To make the design clearer, split `IDEInspectionCallbacks`.

rdar://105120332

ahoppen committed:
Split IDEInspectionCallbacks into CodeCompletionCallbacks and DoneParsingCallback

Cursor info only cares about the `doneParsing` callback and not about all the `complete` functions that are now defined in `CodeCompletionCallbacks`. To make the design clearer, split `IDEInspectionCallbacks`.

rdar://105120332

rintaro reviewed (approved):
Thank you for doing this!

ahoppen committed:
[Parse] Add curly braces

Maybe this makes VSCode happy and resolves a Windows-only compilation failure `error C2059: syntax error: ')'`

ahoppen commented:
@swift-ci Please test Windows

compnerd commented on lib/Parse/ParseExpr.cpp:
```diff
@@ -699,8 +699,9 @@ ParserResult<Expr> Parser::parseExprKeyPath() {
         Context, backslashLoc, rootResult.getPtrOrNull(), CC, hasLeadingDot);
-    if (IDECallbacks)
-      IDECallbacks->completeExprKeyPath(keypath, DotLoc);
+    if (CodeCompletionCallbacks) {
+      CodeCompletionCallbacks->completeExprKeyPath(keypath, DotLoc);
+    }
```
```suggestion
    if (this->CodeCompletionCallbacks)
      this->CodeCompletionCallbacks->completeExprKeyPath(keypath, DotLoc);
```

compnerd commented on lib/Parse/ParseExpr.cpp:
```diff
@@ -3135,8 +3138,10 @@ ParserStatus Parser::parseExprList(tok leftTok, tok rightTok,
       auto CCExpr = new (Context) CodeCompletionExpr(Tok.getLoc());
-      if (IDECallbacks)
-        IDECallbacks->completeCallArg(CCExpr, PreviousLoc == leftLoc);
+      if (CodeCompletionCallbacks) {
+        CodeCompletionCallbacks->completeCallArg(CCExpr,
+                                                 PreviousLoc == leftLoc);
+      }
```
```suggestion
      if (this->CodeCompletionCallbacks)
        this->CodeCompletionCallbacks->completeCallArg(CCExpr,
                                                       PreviousLoc == leftLoc);
```

compnerd reviewed (commented):
This seems to be a bug in the C++ parser in MSVC from VS2019.  VS2022 works properly.  There are a few issues to iron out with the migration on the CI side that I hope that we can get through once @shahmishal returns.  As a temporary measure, please accept my apologies with this workaround.

compnerd committed:
Apply suggestions for VS2019 workaround

The MSVC 2019 14.29 toolset seems to not properly handle the member variable reference in this case.  Workaround this by explicitly using the `this` identifier to indicate that we would like to access the member variable.  This is repaired in newer toolsets, however, the CI is currently using the older toolset.

compnerd commented:
@swift-ci please test

compnerd commented on lib/Parse/ParseExpr.cpp:
```diff
@@ -734,8 +734,9 @@ ParserResult<Expr> Parser::parseExprKeyPathObjC() {
 
-    if (IDECallbacks)
-      IDECallbacks->completeExprKeyPath(expr, DotLoc);
+    if (CodeCompletionCallbacks) {
+      CodeCompletionCallbacks->completeExprKeyPath(expr, DotLoc);
+    }
```
```suggestion
    if (this->CodeCompletionCallbacks)
      this->CodeCompletionCallbacks->completeExprKeyPath(expr, DotLoc);
```

compnerd commented on lib/Parse/ParsePattern.cpp:
```diff
@@ -245,8 +245,9 @@ Parser::parseParameterClause(SourceLoc &leftParenLoc,
       if (AttrStatus.hasCodeCompletion()) {
-        if (IDECallbacks)
-          IDECallbacks->setAttrTargetDeclKind(DeclKind::Param);
+        if (CodeCompletionCallbacks) {
+          CodeCompletionCallbacks->setAttrTargetDeclKind(DeclKind::Param);
+        }
```
```suggestion
        if (this->CodeCompletionCallbacks)
          this->CodeCompletionCallbacks->setAttrTargetDeclKind(DeclKind::Param);
```

compnerd commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -9375,10 +9379,11 @@ Parser::parseDeclPrecedenceGroup(ParseDeclOptions flags,
     if (Tok.is(tok::code_complete)) {
-      if (IDECallbacks)
-        IDECallbacks->completeInPrecedenceGroup(SK);
+      if (CodeCompletionCallbacks) {
+        CodeCompletionCallbacks->completeInPrecedenceGroup(SK);
+      }
```
```suggestion
      if (this->CodeCompletionCallbacks)
        this->CodeCompletionCallbacks->completeInPrecedenceGroup(SK);
```

compnerd committed:
Apply suggestions from MSVC workaround

compnerd commented:
@swift-ci please test Windows platform
